src/cell_list.h

```diff
@@ -0,0 +1,197 @@
+ * It is used to optimize lookup of nearby entities.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list
+{
+    public:
+        cell_list() = default;
+        void add( const point &p, T &value ) {
+            std::pair<point, T *> val( p, &value );
+            cells[p.y / CELL_SIZE][p.x / CELL_SIZE].push_back( val );
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
+            return cell_list_range<T, CELL_SIZE, MAXX, MAXY>( *this, p, distance );
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        std::vector<std::pair<point, T *>> cells[MAXY / CELL_SIZE][MAXX / CELL_SIZE];
+};
+/*
+ * A cell list range handles iteration over elements in the Cell List near some origin point in
+ * nearest-first order and terminates at some specified threshold.
+ * It does this by visiting one "rank" of cells at a time.
+ * First it visits the cell the origin is in, determining the distance to the nearest cell wall,
+ * and iterating over every entity closer than that distance. This guarantees that there is no
+ * closer entity than one found via this process. Once that list is exhausted, it loads one row of
+ * cells around the origin cell and repeating the above process.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range
+{
+    public:
+        cell_list_range( const cell_list< T, CELL_SIZE, MAXX, MAXY > &list, const point &p,
+                         int distance ) : list( &list ), origin( p ), distance( distance ) {
+            origin_cell = {p.x / CELL_SIZE, p.y / CELL_SIZE};
+        }
+        using iterator = cell_list_iterator<T, CELL_SIZE, MAXX, MAXY>;
+        iterator begin() const {
+            return iterator( this );
+        }
+        iterator end() const {
+            return iterator();
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        const cell_list<T, CELL_SIZE, MAXX, MAXY> *list;
+        const point origin;
+        point origin_cell; // = origin / CELL_SIZE
+        const int distance;
+};
+
+
+/*
+ * cell_list_iterator is a forward iterator over the underlying cell_list_range adaptor.
+ * It iterates from nearest entries to most distant entries, and when it reaches
+ * the specified range limit it halts.
+ * The intended use case is to be invoked in a iterator-based for or while loop with
+ * early termination when a valid match is found.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY>
+class cell_list_iterator
+{
+    public:
+        // The default constructor creates an end iterator.
+        cell_list_iterator() : range( nullptr ) {};
+        cell_list_iterator( const cell_list_range< T, CELL_SIZE, MAXX, MAXY > *range ) : range( range ) {
+            for( int i = 0; i < range->distance; ++i ) {
+                sorted_elements.push_back( std::vector<std::pair<point, T * >>() );
```

Comment:
This isn't done, and right here is the reason, this is throwing a bad_alloc exception and I'm at my wits end because I have no idea why.